### PR TITLE
Fix memory issues in remaining RObject methods

### PR DESCRIPTION
**Pull request by lionel-**
I moved the `keep()` calls into `RObject`'s base constructor. This is temporary, in the future `keep()` will only be called from the channel dispatcher.

### CODE REVIEW

File: src/webR/robj-worker.ts [hunk 1]
```diff
@@ -628,33 +639,47 @@ export class RList extends RObject {
...
   }
 }
 
 export class RString extends RObject {
+  // Unlike symbols, strings are not cached and must thus be protected
+  constructor(x: string) {
+    if (typeof x !== 'string') {
+      super(x);
+      return;
+    }
```
Comment (georgestagg) [inline src/webR/robj-worker.ts:672]: The TypeScript signature claims `x` is always `string` here. Should it be something like the following?
```suggestion
  constructor(x: RObject | string) {
    if (isRObject(x)) {
      super(x);
      return;
    }
```

File: src/webR/utils-r.ts [hunk 2]
```diff
@@ -13,6 +13,28 @@ export function protectInc<T extends RHandle>(x: T, prot: { n: number }): T {
...
 
+export function protectWithIndex(x: RHandle): { loc: number; ptr: RPtr } {
+  // Integer size hardcoded to 4 bytes. This is fine but is there a
+  // way to call sizeof?
+  const pLoc = Module._malloc(4);
+  const loc = Module.getValue(pLoc, 'i32');
+
+  Module._R_ProtectWithIndex(handlePtr(x), pLoc);
+
+  return { loc: loc, ptr: pLoc };
```
Comment (georgestagg) [inline src/webR/utils-r.ts:24]: I'm not quite sure what's happening here. If the location pointed to by `pLoc` is populated during `_R_ProtectWithIndex`, should the `Module.getValue()` line come after?

**Review by georgestagg — APPROVED** [review]
This looks really nice, thank you!

I think the use of your memory management helper functions and `new RObject(foo)` where foo is an `RObject` really helps tidy up the code a lot.

I've added a few minor comments but nothing serious.

File: src/webR/robj-worker.ts (hunk 1 above)
Comment (lionel-) [reply src/webR/robj-worker.ts:672]: Good catch.

File: src/webR/utils-r.ts (hunk 2 above)
Comment (lionel-) [reply src/webR/utils-r.ts:24]: aha yes it should! I was wondering why `loc` would cause a crash!

**Pull request merged**